pkg/ebpf/controlplane/controller.go

```diff
@@ -24,103 +26,155 @@ const (
+				probes.RawTracepoint,
+				"cgroup:cgroup_rmdir",
+				"cgroup_rmdir_signal",
+			),
+		},
+	)
 
 	return p, nil
 }
 
-func (p *Controller) Attach() error {
-	err := p.probeGroup.Attach(cgroupMkdirControlProbe)
-	if err != nil {
-		return fmt.Errorf("failed to attach cgroup_mkdir probe in control plane: %v", err)
+func (ctrl *Controller) Attach() error {
+	controlProbes := []probes.Handle{
+		cgroupMkdirControlProbe,
+		cgroupRmdirControlProbe,
 	}
-	err = p.probeGroup.Attach(cgroupRmdirControlProbe)
-	if err != nil {
-		return fmt.Errorf("failed to attach cgroup_rmdir probe in control plane: %v", err)
+
+	// Attach the control probes
+	for _, probeHandle := range controlProbes {
+		err := ctrl.controlProbeGroup.Attach(probeHandle)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to attach control probe (program: %v): %v",
+				ctrl.controlProbeGroup.GetProgramNameByHandle(probeHandle),
+				err,
+			)
+		}
 	}
+
 	return nil
 }
 
-func (p *Controller) Start() error {
-	p.signalBuffer.Poll(pollTimeout)
+func (ctrl *Controller) Start() error {
+	ctrl.signalBuffer.Poll(pollTimeout)
 	return nil
 }
 
-func (p *Controller) Run(ctx context.Context) {
-	p.ctx = ctx
+func (ctrl *Controller) Run(ctx context.Context) {
+	ctrl.ctx = ctx
 	for {
 		select {
-		case signalData := <-p.signalChan:
+		case signalData := <-ctrl.signalChan:
 			signal := signal{}
 			err := signal.Unmarshal(signalData)
 			if err != nil {
 				logger.Errorw("error unmarshaling signal ebpf buffer", "error", err)
 				continue
 			}
-			err = p.processSignal(signal)
+			err = ctrl.processSignal(signal)
 			if err != nil {
 				logger.Errorw("error processing control plane signal", "error", err)
 			}
-		case lost := <-p.lostSignalChan:
+		case lost := <-ctrl.lostSignalChan:
 			logger.Warnw(fmt.Sprintf("Lost %d control plane signals", lost))
-		case <-p.ctx.Done():
+		case <-ctrl.ctx.Done():
 			return
 		}
 	}
 }
 
-func (p *Controller) Stop() error {
-	p.signalBuffer.Stop()
-	return p.probeGroup.DetachAll()
+func (ctrl *Controller) Stop() error {
+	ctrl.signalBuffer.Stop()
+	return ctrl.controlProbeGroup.DetachAll()
 }
 
-func (p *Controller) processSignal(signal signal) error {
+func (ctrl *Controller) processSignal(signal signal) error {
 	switch signal.eventID {
 	case events.CgroupMkdir:
-		return p.processCgroupMkdir(signal.args)
+		return ctrl.processCgroupMkdir(signal.args)
 	case events.CgroupRmdir:
-		return p.processCgroupRmdir(signal.args)
+		return ctrl.processCgroupRmdir(signal.args)
+	case events.SchedProcessFork:
+		return ctrl.processSchedProcessFork(signal.args)
+	case events.SchedProcessExec:
+		return ctrl.processSchedProcessExec(signal.args)
+	case events.SchedProcessExit:
+		return ctrl.processSchedProcessExit(signal.args)
 	}
 	return nil
 }
 
-func (p *Controller) processCgroupMkdir(args []trace.Argument) error {
+//
+// TODO:
+//
+// If we agree that all signal events are just regular events without the event_context_t prefixing
```

Comment:
Good idea if we go for defining the signals explicitly.
We could even go for something more primitive than the signal definition like you suggest and explicitly declare a struct per signal (and then the ebpf program can be a dumb buffer submission which we decode like an event).

Reply:
I'll possible open an issue for that after this is finished.

Reply:
https://github.com/aquasecurity/tracee/issues/3477